File: src/main/java/com/banryeokkurumi/payment/DemoPaymentProvider.java
package com.banryeokkurumi.payment;

import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public interface DemoPaymentProvider {
    CompletableFuture<Result> authorize(UUID orderId, long amount, String idempotencyKey, Scenario scenario);
    Result query(UUID orderId, Scenario scenario);
    Result cancel(UUID orderId, Scenario scenario);

    enum Scenario { SUCCESS, DECLINED, TIMEOUT, CONNECTION_FAILURE }
    enum Status { SUCCEEDED, DECLINED, CANCELLED }
    record Result(Status status, String transactionId, String reason) {}
}

File: src/main/kotlin/com/banryeokkurumi/payment/DemoPaymentProvider.kt
package com.banryeokkurumi.payment

import java.util.UUID
import java.util.concurrent.CompletableFuture

interface DemoPaymentProvider {
    fun authorize(orderId: UUID, amount: Long, idempotencyKey: String, scenario: Scenario): CompletableFuture<Result>

    fun query(orderId: UUID, scenario: Scenario): Result

    fun cancel(orderId: UUID, scenario: Scenario): Result

    enum class Scenario { SUCCESS, DECLINED, TIMEOUT, CONNECTION_FAILURE }

    enum class Status { SUCCEEDED, DECLINED, CANCELLED }

    @JvmRecord
    data class Result(
        val status: Status,
        val transactionId: String?,
        val reason: String?,
    )
}
